INT-3617: ZK Leader Event Processing

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3617

Start/stop `SmartLifecyle` beans with leader election/revocation.

**needs my PRs to spring-cloud-cluster:** https://github.com/spring-cloud/spring-cloud-cluster/pulls

(local mvn package install)

garyrussell committed:
INT-3617: ZK Leader Event Processing

JIRA: https://jira.spring.io/browse/INT-3617

Start/stop `SmartLifecyle` beans with leader election/revocation.

ZK Namespace Support

Add Annotation Support

INT-3617: Polishing

Add `SmartLifecycle` to listener parser.

INT-3617: Polishing; PR Comments

Revert SmartLifecycleRoleController

Remove reflection.

Remove Dependence on spring-cloud-cluster

Temporarily move the relevant classes here.

garyrussell commented:
Added namespace, annotation support; rebased, squashed. Docs will be covered by INT-3747.

garyrussell commented:
Added `SmartLifecycle` to the leader listener.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/annotation/Role.java:
```diff
@@ -0,0 +1,44 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+@Documented
+public @interface Role {
```
Unfortunately we already have a `Role` annotation in Spring and it really can be placed on the `@Bean` method...

garyrussell replied:
That `@Role` is unlikely to be used by user code and if needed one could be fully qualified.

Or, if you prefer, we could consider renaming to `@LeadershipRole`.

artembilan replied:
Since it reflects `AbstractLeaderEvent.getRole()`, it really seems for me like `@LeadershipRole` is a good name.
Or just `@Leadership`. Even if it is a `role`from the `Candidate` it means that our lifecycles are "guided"...
Need to think more, or just leave as `@Role` and make other to live with that! :smile:

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/SmartLifecycleRoleController.java:
```diff
@@ -0,0 +1,194 @@
+		if (event instanceof OnGrantedEvent) {
+			startLifecyclesInRole(event.getRole());
+		}
+		else if (event instanceof OnRevokedEvent) {
```
Looks like these are only a reason for `s-c-c-c` dependency. So, we should rework it to the class names and reflection for the getter.
Otherwise I'm sure that dependency can be `optional` and this `SmartLifecycleRoleController` can be registered only if the first one is in CP.

garyrussell replied:
:+1:

garyrussell replied:
Hmmm - I pushed the polish with this change but we still have a dependency for the generic type of the `ApplicationListener`; otherwise we'll have to accept all events and discard those we aren't interested in.

OTOH, I think we can still have an optional dependency even with hard class names because this bean is only created if they add a `role` to an endpoint.

So I think it can be reverted. WDYT?

artembilan replied:
:+1: But leave that `optional`, please.
Sorry for the noise. I just worry about circular dependency between SI and SC...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractChannelAdapterParser.java:
```diff
@@ -73,6 +73,13 @@ protected final AbstractBeanDefinition parseInternal(Element element, ParserCont
 		if (StringUtils.hasText(phase)) {
 			propertyValues.add("phase", new TypedStringValue(phase));
 		}
+		String role = element.getAttribute("role");
```
I don't see the same for inter-flow endpoints - `AbstractConsumerEndpointParser`.
As well as in the XML test below.
Although you cover that with `@Role`.

Any comments ?

artembilan commented on spring-integration-zookeeper/src/main/resources/META-INF/spring.tooling:
```diff
@@ -0,0 +1,4 @@
+# Tooling related information for the integration Zookeeper namespace
+http\://www.springframework.org/schema/integration/zookeeper@name=integration Zookeeper Namespace
+http\://www.springframework.org/schema/integration/zookeeper@prefix=int-zk
+http\://www.springframework.org/schema/integration/zookeeper@icon=org/springframework/integration/config/xml/spring-integration-zookeeper.gif
```
How have you drown the icon? 
I need a new one for Stomp and seems for me need to fix for WebSocket.
Thanks

garyrussell replied:
I didn't draw a new one, I just copied the generic one from mqtt.

I remember drawing one for ip years ago.

artembilan commented on spring-integration-zookeeper/src/test/java/org/springframework/integration/zookeeper/config/xml/ZookeeperParserTests.java:
```diff
@@ -0,0 +1,73 @@
+
+		@Bean
+		public CuratorFramework client() throws Exception {
+			return createNewClient();
```
Cool trick! :smile: :+1:

garyrussell commented:
Pushed

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/SmartLifecycleRoleController.java:
```diff
@@ -0,0 +1,194 @@
+			addLifecycleToRole(role, lifecycle);
+		}
+		for (List<SmartLifecycle> smarts : this.lifecycles.values()) {
+			Collections.sort(smarts, new Comparator<SmartLifecycle>() {
```
This sort does not make sense here, because we can do `addLifecycleToRole` afterwards.
So, it has to be moved to the `startLifecyclesInRole`

garyrussell replied:
:+1:

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/SmartLifecycleRoleController.java:
```diff
@@ -0,0 +1,194 @@
+	public SmartLifecycleRoleController(List<String> roles, List<SmartLifecycle> lifecycles) {
+		Assert.notNull(roles, "'roles' cannot be null");
+		Assert.notNull(lifecycles, "'lifecycles' cannot be null");
+		Assert.isTrue(roles.size() == lifecycles.size(), "'roles' and 'lifecycles' must be the same lenght");
```
I don't understand this restriction.
We really can have several `Lifecycles` for the same `role`...
Why just don't ask `MultiValueMap` from the constructor?

garyrussell replied:
It's just a lot less work in the parser to add role and SL as separate arguments; and we dont't have `ManagedMultiValueMap`.

This check is to make sure there is a one-to-one correspondence.

But, I suppose a MVM might be easier for JavaConfig; so we could add a second ctor.

artembilan commented on spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/config/xml/LeaderListenerParser.java:
```diff
@@ -0,0 +1,60 @@
+	protected AbstractBeanDefinition parseInternal(Element element, ParserContext parserContext) {
+		BeanDefinitionBuilder candidateBuilder = BeanDefinitionBuilder.genericBeanDefinition(DefaultCandidate.class);
+		candidateBuilder.addConstructorArgValue(UUID.randomUUID().toString());
+		candidateBuilder.addConstructorArgValue(element.getAttribute("role"));
```
I see that Spring-Cloud supports only the single `role` per candidate, but should we really be restricted as well?
I mean comma-separate value for the `role` attr and array value for the `@Role`.
I don't understand the promise of this feature yet. 
So, let me read your second (docs) PR on the matter!

garyrussell replied:
I think we should stick with a single role for now; we can always add more later if users request it.

garyrussell commented:
Removed dependency on s-c-c; PR Comments; rebased, squashed, pushed.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/context/IntegrationContextUtils.java:
```diff
@@ -77,6 +77,8 @@
 	public static final String TO_STRING_FRIENDLY_JSON_NODE_TO_STRING_CONVERTER_BEAN_NAME =
 			"toStringFriendlyJsonNodeToStringConverter";
 
+	public static final String INTEGRATION_LIFECYCLE_ROLE_CONTROLLER = "lifecycleRoleController";
```
Wouldn't `integrationLifecycleRoleController` be better name?
At least for those cases when some target application has done something with similar name...

garyrussell replied:
:+1:

artembilan commented on spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,227 @@
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		start();
```
This one doesn't look as good solution.
If you worry that `applicationContext` doesn't start your `Lifecycle`, just rework it to the `SmartLifecycle`.
Otherwise this implementation does not make sense to me...

Although I see that all these classes are copies from those in the SC. So, the fix must be done there after our M2 already. And we will come back to the dependency on the RC1 dead line

garyrussell replied:
Yes, it was copied temporarily; but please follow [this issue](https://github.com/spring-cloud/spring-cloud-cluster/issues/16).

I didn't notice this; I will fix it here.

artembilan replied:
OK. I'm waiting for the fix.
Thanks

garyrussell replied:
I copied the wrong one...

https://github.com/spring-cloud/spring-cloud-cluster/pull/14

artembilan commented:
Merging if `testAll api reference` is SUCCESSFUL...

artembilan commented on spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,227 @@
+import org.springframework.util.StringUtils;
+
+/**
+ * Bootstrap leadership {@link org.springframework.cloud.cluster.leader.Candidate candidates}
```
Bad link. you can use here the link to your local copy

artembilan commented:
So, tests are fine. Just need all those last fixes to merge.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java:
```diff
@@ -0,0 +1,214 @@
+			addLazyLifecycles();
+		}
+		List<SmartLifecycle> lifecycles = new ArrayList<SmartLifecycle>(this.lifecycles.get(role));
+		if (lifecycles != null) {
```
`!lifecycles.isEmpty()`?
Because `new ArrayList` always produces a new instance.
From other side: does it really make sense to have a copy for the the `MultiMap.get()` result ?

garyrussell replied:
We have to make a copy because of the sort

artembilan commented on spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/leader/LeaderInitiator.java:
```diff
@@ -0,0 +1,267 @@
+	private volatile boolean autoStartup = true;
+
+	/**
+	 * @See SmartLifecycle
```
Bad `@See`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java:
```diff
@@ -0,0 +1,214 @@
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		List<SmartLifecycle> lifecycles = new ArrayList<SmartLifecycle>(this.lifecycles.get(role));
```
It doesn't related to our issue, but if there is no one component in role we end up with NPE from `ArrayList` ctor.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java:
```diff
@@ -0,0 +1,214 @@
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		List<SmartLifecycle> lifecycles = new ArrayList<SmartLifecycle>(this.lifecycles.get(role));
```
The same NPE issue is here

garyrussell committed:
Polishing; PR Comments and Fix Test

Test was incorrectly stopping the LeaderInitiator before it was elected.
Set auto-startup="false" in the i-c-a so we wait for its start before
stopping the LeaderInitiator.

garyrussell commented:
Rebased, pushed.

artembilan commented:
OK. Looks like we can live with this solution.
Merging...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/leader/DefaultCandidate.java:
```diff
@@ -0,0 +1,76 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simple {@link org.springframework.cloud.cluster.leader.Candidate} for leadership.
```
One more bad JavaDoc link.
Fixing on merge...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/814698fbb9e9043930b6101fb9ba80e08f2b935d
